src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php

```diff
@@ -11,55 +11,85 @@
 
 namespace Symfony\Component\Form\Extension\DependencyInjection;
 
+use Psr\Container\ContainerInterface;
 use Symfony\Component\Form\FormExtensionInterface;
 use Symfony\Component\Form\FormTypeGuesserChain;
 use Symfony\Component\Form\Exception\InvalidArgumentException;
-use Symfony\Component\DependencyInjection\ContainerInterface;
 
 class DependencyInjectionExtension implements FormExtensionInterface
 {
-    private $container;
     private $typeServiceIds;
-    private $typeExtensionServiceIds;
     private $guesserServiceIds;
     private $guesser;
     private $guesserLoaded = false;
-
-    public function __construct(ContainerInterface $container, array $typeServiceIds, array $typeExtensionServiceIds, array $guesserServiceIds)
+    private $typeContainer;
+    private $typeExtensionServices;
+    private $guesserServices;
+
+    /**
+     * Constructor.
+     *
+     * @param ContainerInterface $typeContainer
+     * @param iterable[]         $typeExtensionServices
+     * @param array|\Traversable $guesserServices
+     */
+    public function __construct(ContainerInterface $typeContainer, array $typeExtensionServices, $guesserServices, array $guesserServiceIds = null)
     {
-        $this->container = $container;
-        $this->typeServiceIds = $typeServiceIds;
-        $this->typeExtensionServiceIds = $typeExtensionServiceIds;
-        $this->guesserServiceIds = $guesserServiceIds;
+        if (null !== $guesserServiceIds) {
+            @trigger_error(sprintf('Passing four arguments to the %s::__construct() method is deprecated since Symfony 3.3 and will be disallowed in Symfony 4.0. The new constructor only accepts three arguments.', __CLASS__), E_USER_DEPRECATED);
+            $this->guesserServiceIds = $guesserServiceIds;
+            $this->typeServiceIds = $typeExtensionServices;
+        }
+
+        $this->typeContainer = $typeContainer;
+        $this->typeExtensionServices = $typeExtensionServices;
+        $this->guesserServices = $guesserServices;
     }
 
     public function getType($name)
     {
-        if (!isset($this->typeServiceIds[$name])) {
-            throw new InvalidArgumentException(sprintf('The field type "%s" is not registered with the service container.', $name));
+        if (null !== $this->guesserServiceIds) {
```

Comment:
I would add a comment `// BC` to ease readability of the layer, also to help remembering to remove it in 4.0.